lib/spack/spack/environment.py

```diff
@@ -623,6 +628,9 @@ def write_transaction(self):
         """Get a write lock context manager for use in a `with` block."""
         return lk.WriteTransaction(self.txlock, acquire=self._re_read)
 
+    def read_transaction(self):
+        return lk.ReadTransaction(self.txlock)
```

Comment:
Shouldn't the acquire function be `_re_read`?  The lock transactions are built with the assumption that another process may well have modified the environment since you last released the lock.  Note that read will only be called at the beginning of a transaction nest -- transactions assume that while you hold any lock, the state in memory is authoritative.  But without putting `_re_read` here, you risk allowing this:

```python
with self.read_transaction():
    with self.write_transaction():
        # do stuff
```

In that scenario, `_re_read` is never called, as the assumption in the `WriteTransaction` is that the outer `ReadTransaction` would have called it.

I realize that this is only called from `__init__()` to protect a read, but if you change things as discussed above, you could just do this in `__init__()`:

```python
with self.read_transaction():
    pass
```

I think having slightly more awkward code in `__init__()` is worth making the `write_transaction` / `read_transaction` pair safe for potential external callers.

Look at the commit comments on the various lock-related in #14263 for more details on this.  Maybe we better docs on how to use `ReadTransaction` / `WriteTransaction` in `llnl.util.lock` would be helpful... what do you think?

Reply:
`_re_read` doesn't actually set up the `Environment` object if the `manifest_path` doesn't exist (this is to prevent `write_transactions` from overwriting `Environment.__init__` when it is initialized with an `init_file`). Therefore, I don't think I can use `_re_read` as an `acquire_fn` in this case.

I probably wouldn't want to use `_read` as an `acquire_fn` because it doesn't reset the state of the `Environment`.

I intended that `read_transaction` be private for my usage: perhaps I should rename it because I can imagine someone wanting some semantics that you mention in the future, but I need a function that happens to conflict with those semantics. Perhaps I can avoid a function altogether and just create the `ReadTransaction` inside of `__init__`?

Reply:
I think this is the way to go here, unless something starts needing read transactions outside environment.  FWIW, there are some special cases in database where I use some "custom" internal transactions (like for reindex).